flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/generated/ProcessTableRunner.java

```diff
@@ -183,34 +184,53 @@ private void processMethod(RunnableWithException method) throws Exception {
         }
     }
 
+    @SuppressWarnings("unchecked")
     private void moveStateToFunction() throws IOException {
         Arrays.fill(stateCleared, false);
         for (int i = 0; i < stateHandles.length; i++) {
-            final RowData value = stateHandles[i].value();
-            stateToFunction[i] = value;
+            final State stateHandle = stateHandles[i];
+            if (!(stateHandle instanceof ValueState)) {
+                continue;
+            }
+            final ValueState<RowData> valueState = (ValueState<RowData>) stateHandle;
+            final RowData value = valueState.value();
+            valueStateToFunction[i] = value;
         }
     }
 
+    @SuppressWarnings("unchecked")
     private void moveStateFromFunction() throws IOException {
         for (int i = 0; i < stateHandles.length; i++) {
-            final RowData fromFunction = stateFromFunction[i];
-            if (fromFunction == null || isEmpty(fromFunction)) {
-                // Reduce state size
-                stateHandles[i].clear();
+            final State stateHandle = stateHandles[i];
+            if (stateHandle instanceof ValueState) {
+                moveValueStateFromFunction((ValueState<RowData>) stateHandle, i);
             } else {
-                final HashFunction hashCode = stateHashCode[i];
-                final RecordEqualiser equals = stateEquals[i];
-                final RowData toFunction = stateToFunction[i];
-                // Reduce state updates by checking if something has changed
-                if (toFunction == null
-                        || hashCode.hashCode(toFunction) != hashCode.hashCode(fromFunction)
-                        || !equals.equals(toFunction, fromFunction)) {
-                    stateHandles[i].update(fromFunction);
+                if (stateCleared[i]) {
+                    stateHandle.clear();
                 }
             }
         }
     }
 
+    private void moveValueStateFromFunction(ValueState<RowData> valueState, int pos)
+            throws IOException {
+        final RowData fromFunction = valueStateFromFunction[pos];
+        if (fromFunction == null || isEmpty(fromFunction)) {
+            // Reduce state size
+            valueState.clear();
+        } else {
+            final HashFunction hashCode = stateHashCode[pos];
+            final RecordEqualiser equals = stateEquals[pos];
```

Comment:
is it possible to use the equals() method on the functions to assert whether they are equal? If we are not in control as we cannot rely on the equals as they are user written. I am curious on the use of `RecordEqualiser`, is this trying to normalise functions so we can compare them; using the word normalise would seem more intuitive than `RecordEqualiser `to me.  Equaliser implies to me we are looking to make the functions equal - where as we are likely just looking to compare - maybe a RecordComparator might be a better name.

Reply:
We are comparing internal data structures here. i.e. `RowData`. Which might be backed by memory segments. Also SQL semantics need to be considered when comparing nested data. This is why `RecordEqualiser` and `HashFunction` are code generated instances.